File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/keymanagement/KeyManagementServiceTest.java
package com.godaddy.asherah.appencryption.keymanagement;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.time.Instant;
import java.util.function.BiFunction;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.godaddy.asherah.crypto.keys.CryptoKey;

@ExtendWith(MockitoExtension.class)
class KeyManagementServiceTest {
  @Mock
  CryptoKey cryptoKey;
  @Mock
  KeyManagementService keyManagementService;

  @Test
  void testWithDecryptedKey() {
    byte[] keyCipherText = new byte[]{0, 1};
    Instant now = Instant.now();
    boolean revoked = false;
    String expectedResult = "success";
    BiFunction<CryptoKey, Instant, String> actionWithDecryptedKey = (key, instant) -> {
      if (cryptoKey.equals(key) && now.equals(instant)) {
        return expectedResult;
      }
      else {
        return "failure";
      }
    };
    when(keyManagementService.decryptKey(keyCipherText, now, revoked)).thenReturn(cryptoKey);
    when(keyManagementService.withDecryptedKey(any(), any(), anyBoolean(), any())).thenCallRealMethod();

    String actualResult = keyManagementService.withDecryptedKey(keyCipherText, now, revoked, actionWithDecryptedKey);
    assertEquals(expectedResult, actualResult);
  }

}

File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Kms/KeyManagementServiceTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Kms;
using GoDaddy.Asherah.Crypto.Keys;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Kms
{
    [Collection("Logger Fixture collection")]
    public class KeyManagementServiceTest
    {
        private readonly Mock<CryptoKey> cryptoKeyMock;
        private readonly Mock<KeyManagementService> keyManagementServiceMock;

        public KeyManagementServiceTest()
        {
            cryptoKeyMock = new Mock<CryptoKey>();
            keyManagementServiceMock = new Mock<KeyManagementService>();
        }

        [Fact]
        private void TestWithDecryptedKey()
        {
            byte[] keyCipherText = { 0, 1 };
            DateTimeOffset now = DateTimeOffset.UtcNow;
            const bool revoked = false;
            const string expectedResult = "success";

            string ActionWithDecryptedKey(CryptoKey key, DateTimeOffset datetimeOffset)
            {
                if (cryptoKeyMock.Object.Equals(key) && now.Equals(datetimeOffset))
                {
                    return expectedResult;
                }

                return "failure";
            }

            keyManagementServiceMock.Setup(x => x.DecryptKey(keyCipherText, now, revoked))
                .Returns(cryptoKeyMock.Object);
            keyManagementServiceMock.Setup(x =>
                    x.WithDecryptedKey(It.IsAny<byte[]>(), It.IsAny<DateTimeOffset>(), It.IsAny<bool>(), It.IsAny<Func<CryptoKey, DateTimeOffset, string>>()))
                .CallBase();

            string actualResult = keyManagementServiceMock.Object.WithDecryptedKey(keyCipherText, now, revoked, ActionWithDecryptedKey);
            Assert.Equal(expectedResult, actualResult);
        }
    }
}
